2024/02/part1.py

```diff
@@ -0,0 +1,36 @@
+
+
+def read_input(file_name: str) -> list[str]:
+    with open(file_name) as f:
+        return f.readlines()
+
+def is_safe(line: str) -> bool:
+    nums = list(map(int, line.split()))
+    is_increasing = True if nums[0] < nums[1] else False
+    previous = nums[0]
+    for num in nums[1:]:
+        if is_increasing and previous >= num:
+            return False
+        if not is_increasing and previous <= num:
+            return False
+        if not (1 <= abs(num - previous) <= 3):
+            return False
+        previous = num
+    return True
```

Comment:
`previous` と別変数で置かずにループ内で index を用いて扱った方がわかりやすそうです。また `is_increasing` を if 文の分岐に使うのではなく、差分の反転に使うとコードがシンプルになりそうです。こんな感じだとどうでしょうか。

```suggestion
    order = 1 if nums[0] < nums[1] else -1
    for i in range(1, len(nums)):
        diff = (nums[i] - nums[i - 1]) * order
        if not (1 <= diff <= 3):
            return False
    return True
```

Reply:
ありがとうございます。めっちゃすっきりしますね！

今ちょうどhondaさんのPRも読んでるところだったんですが、この`order`、なぜか若干抵抗ある感じを持っていてます。(他の人の意見を聞きたいです)。理由は
- 1, -1の2値しか取らないので、boolで表現したい。(しかし、後から掛け算で活躍しているのでメリットがあるのもわかる)。
  - とはいえ、これくらいよくある範囲なのかもとも思います。
- orderという命名がしっくりきていない。増加列の符号を表しているから`sign_of_sequence`とか思いついたんですがけど、まだなんか違う気もする。

とはいえ、いろいろ書いたんですが、僕のifより読みやすいので、ありがとうございます！

Reply:
`order` という変数名は僕もしっくりきてたわけじゃないので、意見をいただけて助かります。

> 1, -1の2値しか取らないので、boolで表現したい。(しかし、後から掛け算で活躍しているのでメリットがあるのもわかる)。

bool で扱ってこんな感じにするのだとどうでしょうか。個人的に ascending order の場合はそのままにしたいので、`is_decreasing` と意味を反転させた変数名に合わせて変更もしています。

```suggestion
    is_decreasing = True if nums[0] > nums[1] else False
    for i in range(1, len(nums)):
        diff = nums[i] - nums[i - 1]
        if is_decreasing:
            diff *= -1
        if not (1 <= diff <= 3):
            return False
    return True
```

Reply:
あと `is_saf()` は `len(nums) <= 1` の場合にはエラーが出るので、その Validation も入れておいたほうがよさそうですね